File: ADT/stack-queue-pblm/1963.py
#  using stack
#  Tc = O(n)
#  Sc = O(n)
class Solution:
    def minSwaps(self, s: str) -> int:
        st = []
        for ch in s:
            if ch == "[":
                st.append(ch)
            elif st:
                st.pop()
        return (len(st) + 1) // 2


# without using stack
#  Tc = O(n)
#  Sc = O(1)
class Solution:
    def minSwaps(self, s: str) -> int:
        size = 0
        for ch in s:
            if ch == "[":
                size += 1
            elif size != 0:
                size -= 1
        return (size + 1) // 2

File: ADT/stack-queue-pblm/1963.cpp

// using stack
// Tc = O(n)
// Sc = O(n)
class Solution {
public:
    int minSwaps(string s) {
        stack<char> st;

        for (char& ch : s) {
            if (ch == '[')
                st.push(ch);
            else if (!st.empty())
                st.pop();
        }

        return (st.size() + 1) / 2;
    }
};


// without using stack
// Tc = O(n)
// Sc = O(1)
class Solution {
public:
    int minSwaps(string s) {
        int size = 0;

        for (char& ch : s) {
            if (ch == '[')
                size++;
            else if (size != 0)
                size--;
        }

        return (size + 1) / 2;
    }
};